helper/testhelpers/docker/testhelpers.go

```diff
@@ -370,3 +373,239 @@ func copyToContainer(ctx context.Context, dapi *client.Client, containerID, from
 
 	return nil
 }
+
+type RunCmdOpt interface {
+	Apply(cfg *types.ExecConfig) error
+}
+
+type RunCmdUser string
+
+var _ RunCmdOpt = (*RunCmdUser)(nil)
+
+func (u RunCmdUser) Apply(cfg *types.ExecConfig) error {
+	cfg.User = string(u)
+	return nil
+}
+
+func (d *Runner) RunCmdWithOutput(ctx context.Context, container string, cmd []string, opts ...RunCmdOpt) ([]byte, []byte, int, error) {
+	runCfg := types.ExecConfig{
+		AttachStdout: true,
+		AttachStderr: true,
+		Cmd:          cmd,
+	}
+
+	for index, opt := range opts {
+		if err := opt.Apply(&runCfg); err != nil {
+			return nil, nil, -1, fmt.Errorf("error applying option (%d / %v): %v", index, opt, err)
+		}
+	}
+
+	ret, err := d.DockerAPI.ContainerExecCreate(ctx, container, runCfg)
+	if err != nil {
+		return nil, nil, -1, fmt.Errorf("error creating execution environment: %v\ncfg: %v\n", err, runCfg)
+	}
+
+	resp, err := d.DockerAPI.ContainerExecAttach(ctx, ret.ID, types.ExecStartCheck{})
+	if err != nil {
+		return nil, nil, -1, fmt.Errorf("error attaching to command execution: %v\ncfg: %v\nret: %v\n", err, runCfg, ret)
+	}
+	defer resp.Close()
+
+	var stdoutB bytes.Buffer
+	var stderrB bytes.Buffer
+	if _, err := stdcopy.StdCopy(&stdoutB, &stderrB, resp.Reader); err != nil {
+		return nil, nil, -1, fmt.Errorf("error reading command output: %v", err)
+	}
+
+	stdout := stdoutB.Bytes()
+	stderr := stderrB.Bytes()
+
+	// Fetch return code.
+	info, err := d.DockerAPI.ContainerExecInspect(ctx, ret.ID)
+	if err != nil {
+		return stdout, stderr, -1, fmt.Errorf("error reading command exit code: %v", err)
+	}
+
+	return stdout, stderr, info.ExitCode, nil
+}
+
+func (d *Runner) RunCmdInBackground(ctx context.Context, container string, cmd []string, opts ...RunCmdOpt) (string, error) {
+	runCfg := types.ExecConfig{
+		AttachStdout: true,
+		AttachStderr: true,
+		Cmd:          cmd,
+	}
+
+	for index, opt := range opts {
+		if err := opt.Apply(&runCfg); err != nil {
+			return "", fmt.Errorf("error applying option (%d / %v): %v", index, opt, err)
+		}
+	}
+
+	ret, err := d.DockerAPI.ContainerExecCreate(ctx, container, runCfg)
+	if err != nil {
+		return "", fmt.Errorf("error creating execution environment: %v\ncfg: %v\n", err, runCfg)
+	}
+
+	err = d.DockerAPI.ContainerExecStart(ctx, ret.ID, types.ExecStartCheck{})
+	if err != nil {
+		return "", fmt.Errorf("error starting command execution: %v\ncfg: %v\nret: %v\n", err, runCfg, ret)
+	}
+
+	return ret.ID, nil
+}
+
+// Mapping of path->contents
+type PathContents interface {
+	UpdateHeader(header *tar.Header) error
+	Get() ([]byte, error)
+}
+
+type FileContents struct {
+	Data []byte
+	Mode int64
+}
+
+func (b FileContents) UpdateHeader(header *tar.Header) error {
+	header.Mode = b.Mode
+	return nil
+}
+
+func (b FileContents) Get() ([]byte, error) {
+	return b.Data, nil
+}
+
+func PathContentsFromBytes(data []byte) PathContents {
+	return FileContents{
+		Data: data,
+		Mode: 0o644,
+	}
+}
+
+type BuildContext map[string]PathContents
```

Comment:
Might be better to have a slice of pairs instead of a map, for predictable iteration order (though I can't think of any reason why the order of files in a tar stream should matter, I just like as much determinism in my tests as possible, unless there's a good reason otherwise.)

Reply:
Pairs? Do you have a type structure in mind? Would we need to introduce another type here, or better, flatten the path into `PathContents` (perhaps renaming it to `PathContext`)? I think building this then gets harder as you have to use `x = append(x, PathContext{Path, ....})` or `x = append(x, NewPathContextFromString(Path, Value)` rather than `x[path] = ...`.

I [seem to recall](https://utcc.utoronto.ca/~cks/space/blog/unix/ReaddirOrder) disk traversal order is dependent on FS type, so a bit of non-determinism shouldn't matter. 

My 2c.